src/LobbyServer.hs

```diff
@@ -303,3 +295,86 @@ findClient clientName clientList = find ((clientName ==).name) clientList
 -- |Maps over the clients and writes the message to their channel
 messageClients :: LobbyMessage -> [ClientEntry] -> IO ()
 messageClients m cs = mapM_ (\c -> CC.writeChan (lobbyChannel c) m) cs
+
+-- |Called by client to join a chat
+joinChat :: Server ConcurrentClientList -> Server ConcurrentChatList -> String -> Server ()
+joinChat remoteClientList remoteChatList chatName = do
+  sid <- getSessionID
+  concurrentClientList <- remoteClientList
+  concurrentChatList <- remoteChatList
+
+  cs <- liftIO $ CC.readMVar concurrentChatList
+  if isNothing $ chatName `lookup` cs
+    then do
+      newChatChannel <- liftIO CC.newChan
+      liftIO $ CC.modifyMVar_ concurrentChatList $ \chatList ->
+        return $ (chatName, newChatChannel) : chatList
+      addChannelToClient sid concurrentChatList concurrentClientList
+    else
+      addChannelToClient sid concurrentChatList concurrentClientList
+
+
+    where
+      addChannelToClient sid concurrentChatList concurrentClientList = do
+        cs <- liftIO $ CC.readMVar concurrentChatList
+        let chan = fromJust $ chatName `lookup` cs
+        clientChan <- liftIO $ CC.dupChan chan
+        let newChat = (chatName, clientChan)
+        liftIO $ CC.modifyMVar_ concurrentClientList $ \clients -> do
+          let client = fromJust $ lookupClientEntry sid clients
```

Comment:
I wouls say so yes. Since it's checked and added if it doesn't exists in the if-then-else in the lines above.